### ISSUE TITLE
Typing seems correct to me but compiler complains. 

### ISSUE THREAD

**Issue by fxfactorial** [ocsigen/js_of_ocaml#372]
Say we have a binding to an object like so: 

```ocaml
class type browser_window = object

  method loadUrl : Js.js_string Js.t -> unit Js.meth

  method openDevTools : unit -> unit Js.meth

  method webContents : web_contents Js.t Js.readonly_prop

  method on :
    Js.js_string Js.t ->
    (unit -> unit) Js.callback ->
    unit Js.meth

end
```

Where `web_contents` is defined as: 

```ocaml
class type web_contents = object

  method on :
    Js.js_string Js.t ->
    (unit -> unit) Js.callback ->
    unit Js.meth

end
```

And I have code like so 

```ocaml

(* This types correctly *)
let web_con = some_browser##.webContents in
(* This craps out *)
web_con##on (Js.string "something") (Js.wrap_callback (fun () -> print_endline "Called"));
```

I get a error of: 

```shell
Error: This expression has type web_contents Js.t
       but an expression was expected of type (< .. > as 'a) Js.t
       Type web_contents is not compatible with type 'a 
```
Which is odd to me, I did a binding of this kind before without issue. For what its worth, I am creating the `browser_window` as a `Js.constr`.

**Comment by vasilisp** [comment]
Try to parenthesize and comma-separate the arguments:

```ocaml
web_con##on(Js.string "something", Js.wrap_callback (fun () -> print_endline "Called")));
```

The syntax reflects the fact that JS functions cannot be partially applied (unlike OCaml functions).

**Comment by Drup** [comment]
@vasilisp No. He's using the ppx syntax.

**Comment by vasilisp** [comment]
Sorry, ignore me, I am too used to Camlp4 :).

**Comment by fxfactorial** [comment]
Okay then I'm starting to think its a bug with ppx/jsoo.

**Comment by Drup** [comment]
[This](https://bpaste.net/show/ee0743a1228a) types fine for me, on normal dev jsoo, with the command line`ocamlfind ocamlc -i -package "js_of_ocaml.ppx" bla.ml `

**Comment by fxfactorial** [comment]
@Drup 

Here's my usage that does work, 

![screen shot 2015-10-27 at 7 50 21 am](https://cloud.githubusercontent.com/assets/3036816/10761629/7c2262b2-7c7f-11e5-937c-a96d6f37573d.png)

And here's my usage that doesn't work. 

![screen shot 2015-10-27 at 7 51 39 am](https://cloud.githubusercontent.com/assets/3036816/10761649/9ac75aba-7c7f-11e5-8f7d-f9bc83d4ae6a.png)

**Comment by Drup** [comment]
I'm not sure what kind of divination you expect us to do with screenshots of code. :)
Please provide an actual complete piece of code that fails, so that we can reproduce and fix.

**Comment by fxfactorial** [comment]
To get to this point is a lot of dependencies, obviously I would have given a small piece of reproducible code if I could. To get to this point you need.

1) electron installed, `npm install electron-prebuilt -g`
2) nodejs, `opam pin add nodejs https://github.com/fxfactorial/ocaml-nodejs`
3) electron bindings. `opam pin add electron https://github.com/fxfactorial/ocaml-electron`
4) Usage: , 4.1) `git clone https://github.com/fxfactorial/science-hack` 4.2) `make`

The trouble some usage is line `31` in science-hack of `fruitbat_renderer.ml`

**Comment by hhugo** [comment]
change
`method getCurrentWindow : unit -> Browser_window.browser_window Js.t Js.meth`
by
 `method getCurrentWindow : Browser_window.browser_window Js.t Js.meth`

**Comment by fxfactorial** [comment]
@hhugo Same, in any case I thought that methods in OCaml always ran when they were used.

**Comment by hhugo** [comment]
What version of js_of_ocaml are you using ? `js_of_ocaml --version`
Do you use any modified version from https://github.com/the-lambda-church/merlin/issues/449 ?
cc @def-lkb

**Comment by fxfactorial** [comment]
@hhugo I have tried both the opam 2.6 version and @def-lkb 's pin suggestion.

**Comment by fxfactorial** [comment]
@hhugo I think something is wrong with the PPX or js_of_ocaml code because I removed that unit argument as so,

`method getCurrentWindow : Browser_window.browser_window Js.t Js.meth`

and use 
```
let b_win = remote##getCurrentWindow in
b_win##openDevTools ();
```

for which I get this type error, 
```
File "fruitbat_render.ml", line 32, characters 8-13:
Error: This expression has type Browser_window.browser_window
       It has no method openDevTools
make: *** [all] Error 2
```

Where is the `Js.t`?  

This is with @def-lkb 's pin, 
```
js_of_ocaml --version                                                   ⏎
2.6+git-217891d
```

With opam's js_of_ocaml version, which is 2.6, I get that familiar 

```
Error: This expression has type Browser_window.browser_window Js.t
       but an expression was expected of type (< .. > as 'a) Js.t
       Type Browser_window.browser_window is not compatible with type 'a 
```

error.

**Comment by hhugo** [comment]
You should remove the `unit` argument from `openDevTools`

**Comment by fxfactorial** [comment]
Done, got same:

```
Error: This expression has type Browser_window.browser_window Js.t
       but an expression was expected of type (< .. > as 'a) Js.t
       Type Browser_window.browser_window is not compatible with type 'a 
```

**Comment by hhugo** [comment]
I think your issue has to do with type/module aliases.

```
File "fruitbat_render.ml", line 27, characters 71-98:
Error: This expression has type Browser_window.browser_window Js.t
       but an expression was expected of type
         Electron_main.Browser_window.browser_window Js.t
       Type Browser_window.browser_window is not compatible with type
         Electron_main.Browser_window.browser_window =
           < center : unit -> unit Js.meth; close : unit -> unit Js.meth;
             destroy : unit -> unit Js.meth;
.....
```

**Comment by fxfactorial** [comment]
No, I don't think it's that because I have changed the method signature of `getCurrentWindow` to 

```ocaml
  method getCurrentWindow : Electron_main.Browser_window.browser_window Js.t Js.meth
```

And still get the same error, in fact it doesn't even reference the `Electron_main` module at all in the error, its the same as I've posted.

**Comment by hhugo** [comment]
in _oasis file you define `Browser_window` as being an `InternalModules`. This tells oasis not to install cmi for this module. `remote##getCurrentWindow` has type `Browser_window.browser_window Js.t` but the compiler doesn't know details about `Browser_window.browser_window` because there is no cmi installed.

**Comment by fxfactorial** [comment]
Removed all `InternalModules` from the `_oasis`, same error.

**Comment by hhugo** [comment]
Looking at your repos, this issue seems to resolved.
reopen if needed

**Comment by fxfactorial** [comment]
yes, thank you

**Comment by hhugo** [comment]
with https://github.com/ocaml/ocaml/pull/335 (included in 4.03), you should get better errors here.
